arbnode/dataposter/storage/storage.go

```diff
@@ -24,3 +27,117 @@ type QueuedTransaction struct {
+	Data            types.DynamicFeeTx
+	Meta            BatchPosterPosition
+	Sent            bool
+	Created         time.Time // may be earlier than the tx was given to the tx poster
+	NextReplacement time.Time
+}
+
+// This is also for legacy reason. Since Batchposter is in arbnode package,
+// we can't refer to BatchPosterPosition type there even if we export it (that
+// would create cyclic dependency).
+// Ideally we'll factor out Batch Poster from arbnode into separate package
+// and BatchPosterPosition into another separate package as well.
+// For the sake of minimal refactoring, that struct is duplicated here.
+type BatchPosterPosition struct {
+	MessageCount        arbutil.MessageIndex
+	DelayedMessageCount uint64
+	NextSeqNum          uint64
+}
+
+func DecodeLegacyQueuedTransaction(data []byte) (*LegacyQueuedTransaction, error) {
+	var val LegacyQueuedTransaction
+	if err := rlp.DecodeBytes(data, &val); err != nil {
+		return nil, fmt.Errorf("decoding legacy queued transaction: %w", err)
+	}
+	return &val, nil
+}
+
+func LegacyToQueuedTransaction(legacyQT *LegacyQueuedTransaction) (*QueuedTransaction, error) {
+	meta, err := rlp.EncodeToBytes(legacyQT.Meta)
+	if err != nil {
+		return nil, fmt.Errorf("converting legacy to queued transaction: %w", err)
+	}
+	return &QueuedTransaction{
+		FullTx:          legacyQT.FullTx,
+		Data:            legacyQT.Data,
+		Meta:            meta,
+		Sent:            legacyQT.Sent,
+		Created:         legacyQT.Created,
+		NextReplacement: legacyQT.NextReplacement,
+	}, nil
+}
+
+func QueuedTransactionToLegacy(qt *QueuedTransaction) (*LegacyQueuedTransaction, error) {
+	if qt == nil {
+		return nil, nil
+	}
+	var meta BatchPosterPosition
+	if qt.Meta != nil {
+		if err := rlp.DecodeBytes(qt.Meta, &meta); err != nil {
+			return nil, fmt.Errorf("converting queued transaction to legacy: %w", err)
+		}
+	}
+	return &LegacyQueuedTransaction{
+		FullTx:          qt.FullTx,
+		Data:            qt.Data,
+		Meta:            meta,
+		Sent:            qt.Sent,
+		Created:         qt.Created,
+		NextReplacement: qt.NextReplacement,
+	}, nil
+}
+
+type EncoderDecoder struct{}
+
+func (e *EncoderDecoder) Encode(qt *QueuedTransaction) ([]byte, error) {
+	return rlp.EncodeToBytes(qt)
+}
+
+func (e *EncoderDecoder) Decode(data []byte) (*QueuedTransaction, error) {
+	var item QueuedTransaction
+	if err := rlp.DecodeBytes(data, &item); err != nil {
+		return nil, fmt.Errorf("decoding item: %w", err)
+	}
+	return &item, nil
+}
+
+type LegacyEncoderDecoder struct{}
+
+func (e *LegacyEncoderDecoder) Encode(qt *QueuedTransaction) ([]byte, error) {
+	legacyQt, err := QueuedTransactionToLegacy(qt)
+	if err != nil {
+		return nil, fmt.Errorf("encoding legacy item: %w", err)
+	}
+	return rlp.EncodeToBytes(legacyQt)
+}
+
+func (le *LegacyEncoderDecoder) Decode(data []byte) (*QueuedTransaction, error) {
+	val, err := DecodeLegacyQueuedTransaction(data)
+	if err != nil {
+		return nil, fmt.Errorf("decoding legacy item: %w", err)
```

Comment:
I think we need both LegacyEncoderDecoder and EncoderDecoder to support reading the other's format (we might want to make that just a universal function and make the interface only for encoding). In terms of the transition, we'll have 2 steps roughly:
- Move from the old batch poster node version to the new version with legacy encoding enabled:
  - This works because the old poster can read what the new poster writes (because it's legacy format), and the new poster can read what the old poster writes (because it has legacy encoding enabled)
- Enable new format encoding / disable legacy encoding:
  - This doesn't work because the old poster with legacy encoding disabled can't read the new poster's new format queued transactions, and the new poster with legacy encoding enabled can't read the old poster's legacy format queued transactions.

Reply:
Good point, introduced general decode() method that attempts to decode QueuedTransaction, and if that fails then attempts to decode legacyQueuedTransaction and convert.